File: Question/19_Remove_Nth_Node_From_End_of_List.cpp
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* right = head;
        int count = n;
        while(count) {
            right = right->next;
            --count;
        }

        ListNode* left = head;
        ListNode* prev = left;
        while(right) {
            right = right->next;
            prev = left;
            left = left->next;
        }
        if(prev == left) {
            ListNode* temp  = head;
            head = head->next;
            delete temp;
        } else {
            prev->next = left->next;
            delete left;
        }

        return head;
    }
};

File: Question_Go/19_Remove_Nth_Node_From_End_of_List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func removeNthFromEnd(head *ListNode, n int) *ListNode {
    if head == nil {
        return nil
    }

    dummy := ListNode{-1, head}
    prev := &dummy
    cur := &dummy
    for i := 0; i < n; i++ {
        cur = cur.Next
    }

    for cur.Next != nil {
        prev = prev.Next
        cur = cur.Next
    }
    prev.Next = prev.Next.Next
    return dummy.Next
}